### DISCUSSION TITLE
how can I put a literal curly-brace in the message

### DISCUSSION THREAD

**Discussion by ninesunsabiu** [lingui/js-lingui#1894] [Q&A] [answered]
I have a text like this: `tip: use {var} in your comment, then...`
I want to directly display the `{` and `}` symbols to the user, but after `extract` they are recognized as a variable.

I've tried adding `'` but it doesn't seem perfect.

I referred to the method on https://wiki.zimbra.com/wiki/Translations.

**Comment by andrii-bodnar** [comment]
It should work with escaping with the `'` character (reference - https://github.com/lingui/js-lingui/issues/1075)

Test case:
https://github.com/lingui/js-lingui/blob/f15ce1520efcf33fefc160f6f59fa86ead3183a3/packages/core/src/interpolate.test.ts#L34-L38

**Reply by ninesunsabiu** [reply]
<img width="904" alt="image" src="https://github.com/lingui/js-lingui/assets/19882837/f601bfe8-c59f-4347-888c-44f503e5b34e">

after extract

```
#: src/modules/pages/SystemManagement/FormulaManagement/FormulaInfo/Operation/index.tsx:297
msgid "<0><1><2>使用 # 可以调用计算公式</2></1></0><3><4><5>支持的计算因子有：</5><6>'{估算人天}': 用来表示卡片的数值属性或者计算公式</6><7>具体数值</7></4></3><8><9><10>支持的运算符有：</10><11>( ) + - * /</11></9></8>"
msgstr "<0><1><2>使用 # 可以调用计算公式</2></1></0><3><4><5>支持的计算因子有：</5><6>'{估算人天}': 用来表示卡片的数值属性或者计算公式</6><7>具体数值</7></4></3><8><9><10>支持的运算符有：</10><11>( ) + - * /</11></9></8>"
```

and the user interface is

<img width="409" alt="image" src="https://github.com/lingui/js-lingui/assets/19882837/96670509-3570-4cb9-a680-06c5f559e6de">

and I loose the `{估算人天}`


I suspect this has to do with how the JSX macro operates. I'll try again later with the Vanilla JS case.

**Reply by timofei-iatsenko** [reply]
You'd better to put trans into each section separately (in every AntdCol) instead of wrapping big chunk of JSX in it.

**Reply by ninesunsabiu** [reply]
But I think it's a complete piece of content. Would it be less friendly to translators if it were separated?

BTW

<img width="764" alt="image" src="https://github.com/lingui/js-lingui/assets/19882837/0f05b07b-b66a-48c6-b70e-57fa325754c5">


```
#: src/modules/pages/SystemManagement/FormulaManagement/FormulaInfo/Operation/index.tsx:306
msgid "'{估算人天}': 用来表示卡片的数值属性或者计算公式"
msgstr "'{估算人天}': 用来表示卡片的数值属性或者计算公式"
```

and it the same as wrapping big chunk

**Reply by timofei-iatsenko** [reply]
yes, it was not related, just a quick tip. You can change this particualare place to use `t` instead of `<Trans>`, that should fix your issue with escaping

**Reply by ninesunsabiu** [reply]
<img width="957" alt="image" src="https://github.com/lingui/js-lingui/assets/19882837/561cf828-727d-4618-bf73-219a9c00d05c">

<img width="797" alt="image" src="https://github.com/lingui/js-lingui/assets/19882837/1379f69e-2123-4f6f-bb09-287597a423b7">

Oh, It does not work

**Reply by timofei-iatsenko** [reply]
It's not clear from these screenshots how exactly It does not work

**Reply by ninesunsabiu** [reply]
I'll make a minimal a minimal reproduction repo later~

**Reply by ninesunsabiu** [reply]
@thekip 

[ninesunsabiu/lingui-brace-bug: a minimal reproduction for literal brace use case](https://github.com/ninesunsabiu/lingui-brace-bug)

in this project, I tried by two ways
<img width="639" alt="image" src="https://github.com/lingui/js-lingui/assets/19882837/93335818-a390-4554-8beb-1dc6bc8be0d7">

Unfortunately, it does work

<img width="731" alt="image" src="https://github.com/lingui/js-lingui/assets/19882837/c3e7e5cc-6409-4e62-8abb-4fac89a3aa70">

**Reply by ninesunsabiu** [reply]
I run `lingui compile --typescript` to see what `po` file output

and I notice the `'` character is missing

<img width="990" alt="image" src="https://github.com/lingui/js-lingui/assets/19882837/16f50efe-c0f7-45e9-9e6b-8a7a02e1e005">

```
#: src/App.tsx:18
#: src/App.tsx:35
msgid "Hey '{name}'!"
msgstr "!'{eman}' yeH"
```

```
{\"qsNH+W\":\"!{eman} yeH\"}
```

 in `i18n.ts`, it run `compileMessage` with `!{eman} yeH`. and `!{eman} yeH` does not includes `'` character 

<img width="678" alt="image" src="https://github.com/lingui/js-lingui/assets/19882837/394af9de-9165-44f4-b3f2-28e6fcffee1e">

maybe, there is something wrong in `compile` process?

**Reply by timofei-iatsenko** [reply]
Well, i see the bug. The message gets double compiled in dev mode and curly braces got processed second time. 

> and I notice the ' character is missing

That is expected. It's compiled to a string. When there is interpolation there would be an array with 3 entries, instead one string. This an internal lingui catalog representation.

**Reply by ninesunsabiu** [reply]
Why message should compile in dev mode ?  Is it for handling scenarios like `t({ message: 'something {theVar}', id: 'abcd' })` that haven't been extracted in development mode?

change the code to 

```diff
-    let translation = messageForId || message || id

+    let translation = messageForId || (() => {
+        return process.env.NODE_ENV !== "production" && isString(message) ? compileMessage(message) : undefined
+    })() || id
 
-    if (process.env.NODE_ENV !== "production") {
-      translation = isString(translation)
-        ? compileMessage(translation)
-        : translation
-    }

```

only compile `options#message` in dev mode. 

Will this change solve the problem?

@thekip

**Reply by timofei-iatsenko** [reply]
Yes, that's exactly how i wanted to solve this issue. You need to bypass compilation for `messageForId` but still compile `message || id`. But this change should also be covered by tests and i also could not predict how it can affect other users (what about if some loading not compiled catalogs during dev? This could break the flow for them, even though it's a misuse of a lib)

**Reply by ninesunsabiu** [reply]
I tried modifying it according to the above method. Unfortunately, I found that many test cases in the unit tests rely on the behavior of compiling `messageForId`.

for example:

https://github.com/lingui/js-lingui/blob/69c1c235c2fe6264559ac2ea4a7fc2b51b0633b6/packages/core/src/i18n.test.ts#L189-L203

in real word, the value of `messages#['My name is {name}']` is `["Je m'appelle ",["name"]]` (compile from po file by cli)

should I change these test case ?

**Reply by timofei-iatsenko** [reply]
Yes, it's better to change to the compiled strings. You also can change the ID to something short to avoid misunderstanding in the tests: 

```js
 "user.name": ["Je m'appelle ",["name"]], 
```

**Comment by timofei-iatsenko** [comment, marked as answer]
i created an issue https://github.com/lingui/js-lingui/issues/1901

**Reply by ninesunsabiu** [reply]
thx~
